metagraph/src/graph/representation/succinct/dbg_succinct_cached.hpp

```diff
@@ -0,0 +1,235 @@
+#ifndef __DBG_SUCCINCT_CACHED__
+#define __DBG_SUCCINCT_CACHED__
+
+#include <cache.hpp>
+#include <lru_cache_policy.hpp>
+
+#include "boss.hpp"
+#include "dbg_succinct.hpp"
+#include "common/seq_tools/reverse_complement.hpp"
+
+namespace mtg {
+namespace graph {
+
+class DBGSuccinctCached : public DeBruijnGraph {
+  public:
+    typedef boss::BOSS::edge_index edge_index;
+    typedef boss::BOSS::TAlphabet TAlphabet;
+
+    template <typename Key, typename Value>
+    using LRUCache = caches::fixed_sized_cache<Key, Value, caches::LRUCachePolicy<Key>>;
+
+    typedef LRUCache<node_index, std::string> DecodedCache;
+    typedef LRUCache<edge_index, edge_index> BwdFirstCache;
+
+    DBGSuccinctCached(const DBGSuccinct &dbg, size_t cache_size = 1000)
+          : graph_(dbg), boss_(graph_.get_boss()),
+            bwd_first_cache_(cache_size),
+            decoded_cache_(cache_size * dbg.alphabet().size()) {}
+
+    DBGSuccinctCached(const DeBruijnGraph &graph, size_t cache_size = 1000)
+          : DBGSuccinctCached(dynamic_cast<const DBGSuccinct&>(graph), cache_size) {}
+
+    virtual const DeBruijnGraph& get_base_graph() const override final { return graph_.get_base_graph(); }
+
+    const DBGSuccinct& get_dbg_succ() const { return graph_; }
+
+    inline void put_decoded_node(node_index node, const std::string &seq) const {
+        decoded_cache_.Put(node, seq);
+    }
```

Comment:
I think if we make this class, the cache should be efficient.

1) we should store k-mers compressed (probably as `KMer`?) and not std::string
2) we should use a more efficient cache library (a good hash table, not std::unordered_map which is used there at the moment) and std::deque instead of std::list.

Though we can postpone 2) for later

Reply:
Storing k-mers in compressed form improved the overall alignment performance by 15%.

For now, I've added a TODO for the second point. I'll benchmark this thing to see if the cache operations are a significant bottleneck.

Reply:
I suggested this mainly to reduce the memory footprint, so you can make the cache size much larger.

Reply:
That's one thing, but an additional benefit of having compressed k-mers is that it allowed for a sequence of adjacent k-mers to be efficiently generated and cached through small copies and bitwise operations, rather than copying and concatenating strings.